src/api/ucc.h

```diff
@@ -261,27 +261,28 @@ typedef enum {
 /**
  *  @ingroup UCC_COLLECTIVES
  *
- *  @brief The reduction wrapper provides a method to map custom user types to higher level
- *  programming model datatypes
+ *  @brief The reduction wrapper provides an interface for the UCC library to invoke
+ *         user-defined custom reduction callback
  *
  *  @param [in]  invec          The input elements to be reduced by the user function
  *  @param [in]  inoutvec       The input elements to be reduced and output of the reduction
  *  @param [in]  count          The number of elements of type "dtype" to be reduced
- *  @param [in]  dtype          Datatype passed to the reduction operation
+ *  @param [in]  dtype          Datatype specified in the coll_args
+ *  @param [in]  custom_op      A pointer to the user defined reduction passed to the coll_args as custom_reduction_op
  *
  *
  *  @parblock
  *
  *  @b Description
  *
- *  This function is called by the UCC library before calling the user-defined reduction.
- *  Hence, the signature of this function is same as ucc_userdefined_reductions_op_t.
- *  It maps the custom user types to higher level programming model datatypes (such as MPI datatypes)
+ *  This function is called by the UCC library when it needs to perform a non-standard
+ *  user-defined reduction operaion during allreduce/reduce collective.
  *
  *  @endparblock
  */
-typedef void(*ucc_reduction_dtype_mapper_t)(void *invec, void *inoutvec,
-                                            ucc_count_t *count, ucc_datatype_t dtype);
+typedef void(*ucc_reduction_wrapper_t)(void *invec, void *inoutvec,
+                                       ucc_count_t *count, ucc_datatype_t dtype,
+                                       void *custom_reduction_op);
```

Comment:
@vspetrov You want datatype to be void* instead of ucc_datatype_t for two reasons 1) OMPI is operating under ompi datatype and 2) future proof for supporting user-defined datatype (opaque). For (2) void* datatype is not sufficient but necessary.

Reply:
@manjugv  What will i pass to void* argument when i call the ucc_reduction_wrapper callback inside UCC? User provides us with ucc_datatype_t. Will i just cast like this:
```
ucc_knomial_allreduce (...) {
    ... some code ...
    ucc_reduction_wrapper (sbuf, rbuf, &count, (void*)&ucc_datatype, custom_op);
}
```
And how in this case OMPI will treat this void* to implement the callback? We decided that user does NOT pass "ompi_datatype_t" into UCC right now. We don't a place for that. So, i HAVE to use ucc_datatype_t. I suggested the solution where all the information is "packed" into ucc_datatype_t (both pointer and ucc_datatype) you said you didn't like it. "void*" - does not work.

Reply:
@vspetrov Once you call the wrapper, you don't care about ucc datatype, right ? You only care about how to get to ompi datatype type. I modified your code snippet to show this. Let me know, if it looks right? Let’s discuss this in the WG tomorrow.

Reply:
```C
/*
 * When user-defined op and user-defined datatype 
 * ompi/coll/ucc/mca_ucc_ops.c:
 */

typedef void(*ucc_reduction_wrapper_t)(void *invec, void *inoutvec,
                                                                 ucc_count_t *count, void *dtype,  void *op) {
	ompi_op_t *op = (ompi_op_t*)(op);
	ompi_datatype_t dt = (ompi_dt_t*)(dtype); // for pre-defined types - fixed translation table
	ompi_dt_reduce(invec, inoutvec, count, dt, op); //actual invocation for a callback
}


int mca_coll_ucc_allreduce(const void *sbuf, void *rbuf, int count, struct ompi_datatype_t *dtype,
                            struct ompi_op_t *op, struct ompi_communicator_t *comm,
                            mca_coll_base_module_t *module)
{
    mca_coll_ucc_module_t *ucc_module = (mca_coll_ucc_module_t*)module;
    ucc_coll_req_h         req;
    ucc_datatype_t         ucc_dt;
    ucc_reduction_op_t     ucc_op;

    ucc_dt = ompi_dtype_to_ucc_dtype(dtype); //const lookup table for pre-defined types
    ucc_op = ompi_op_to_ucc_op(op); // const lookup table for pre-defined ops
   
    ucc_op = ucc_op_make_user
    ucc_coll_op_args_t coll = {
        .coll_type = UCC_ALLREDUCE,
        .buffer_info = {
            reduce.src_buffer = (void*)sbuf,
            reduce.dst_buffer = rbuf,
            reduce.count      = count,
            reduce.datatype   = ucc_dt,
            reduce.op         = ucc_op,
        },
        if (UCC_OP_USERDEFINED)||(UCC_DT_OPAQUE == ucc_dt) 
        {
            .buffer_info.reduce.custom_reduction_op = (void*)op;
            .user_dtype = (void*)dtype; /* This is a new field */
        }
    };

    COLL_UCC_CHECK(ucc_collective_init(&coll, &req, ucc_module->ucc_team));
    COLL_UCC_CHECK(ucc_collective_post(req));
    COLL_UCC_CHECK(coll_ucc_req_wait(req));
    return OMPI_SUCCESS;
}

In the UCC library .. 

ucc_allreduce()
{

    if (coll_args.buffer_info->reduce.op == UCC_OP_USERDEFINED) {
        reduction_wrapper(inbuf, outbuf, count, user_dtype, custom_reduction_op);
    }
}
```